pkg/server/plugin/keymanager/gcpkms/gcpkms.go

```diff
@@ -0,0 +1,911 @@
+
+				// There was an error in the DestroyCryptoKeyVersion call.
+				// Try to get the CryptoKeyVersion to know the state of the
+				// CryptoKeyVersion and if we need to re-enqueue.
+				cryptoKeyVersion, err := p.kmsClient.GetCryptoKeyVersion(ctx, &kmspb.GetCryptoKeyVersionRequest{
+					Name: cryptoKeyVersionName,
+				})
+				switch status.Code(err) {
+				case codes.NotFound:
+					// Purely defensive. We don't really expect this situation,
+					// because this should have been captured during the
+					// DestroyCryptoKeyVersion call that was just performed.
+					log.Warn("CryptoKeyVersion not found")
+					p.setInactive(ctx, cryptoKey)
+					backoff = backoffMin
+					p.notifyDestroy(err)
+					continue
+				case codes.OK:
+					if cryptoKeyVersion.State != kmspb.CryptoKeyVersion_ENABLED {
+						// Something external to the plugin modified the state
+						// of the CryptoKeyVersion. Do not try to schedule it for
+						// destruction.
+						log.Warn("CryptoKeyVersion is not enabled, will not be scheduled for destruction", "crypto_key_version_state", cryptoKeyVersion.State.String())
+						p.setInactive(ctx, cryptoKey)
+						backoff = backoffMin
+						p.notifyDestroy(err)
+						continue
+					}
+				default:
+					// The GetCryptoKeyVersion call failed. Log this and re-enqueue
+					// the CryptoKey for destruction. Hopefully, this is a
+					// recoverable error.
+					log.Error("Could not get the CryptoKeyVersion while trying to schedule it for destruction")
+				}
+
+				select {
+				case p.scheduleDestroy <- cryptoKey:
+					log.Debug("CryptoKeyVersion re-enqueued for destruction")
+				default:
+					log.Error("Failed to re-enqueue CryptoKeyVersion for destruction")
+				}
+			}
+			p.notifyDestroy(err)
+			backoff = min(backoff*2, backoffMax)
+			p.hooks.clk.Sleep(backoff)
+		}
+	}
+}
+
+// setInactive updates the spire-active label in the specified CryptoKey to
+// indicate that is inactive.
+func (p *Plugin) setInactive(ctx context.Context, cryptoKey *kmspb.CryptoKey) {
+	log := p.log.With(cryptoKeyNameTag, cryptoKey.Name)
+
+	cryptoKey.Labels["spire-active"] = "false"
+	_, err := p.kmsClient.UpdateCryptoKey(ctx, &kmspb.UpdateCryptoKeyRequest{
+		UpdateMask: &fieldmaskpb.FieldMask{
+			Paths: []string{"labels"},
+		},
+		CryptoKey: cryptoKey,
+	})
+	if err != nil {
+		log.Error("Could not update CryptoKey as incactive", reasonTag, err)
+	}
+
+	log.Debug("CryptoKey updated as inactive", cryptoKeyNameTag, cryptoKey.Name)
+}
+
+// setCache sets the cached entries with the provided entries.
+func (p *Plugin) setCache(keyEntries []*keyEntry) {
+	p.entriesMtx.Lock()
+	defer p.entriesMtx.Unlock()
+
+	p.entries = make(map[string]keyEntry)
+
+	for _, e := range keyEntries {
+		p.entries[e.publicKey.Id] = *e
+		p.log.Debug("Cloud KMS key loaded", cryptoKeyVersionNameTag, e.cryptoKeyVersion.Name, algorithmTag, e.cryptoKeyVersion.Algorithm)
```

Comment:
as a general comment, I think log will already have the plugin name as a fields so Cloud KMS is not required as part of the logs.